x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,541 @@
+            public int getPrecedingGapSize(int fragmentNum) {
+                return precedingGapSizes[fragmentNum];
+            }
+
+            public boolean isMatchAll() {
+                return fragments.length == 0 && wildcardStringCount >0 && wildcardCharCount ==0;
+            }
+
+            @Override
+            public int hashCode() {
+                return pattern.hashCode();
+            }
+
+            @Override
+            public boolean equals(Object obj) {
+                PatternStructure other = (PatternStructure) obj;
+                return pattern.equals(other.pattern);
+            }
+            
+            
+        }
+        
+
+        @Override
+        public Query wildcardQuery(String wildcardPattern, RewriteMethod method, QueryShardContext context) {
+            PatternStructure patternStructure = new PatternStructure(wildcardPattern);
+            
+            
+            BooleanQuery.Builder bqBuilder = new BooleanQuery.Builder();
+            
+            for (int i = 0; i < patternStructure.fragments.length; i++) {
+                String fragment = patternStructure.fragments[i];
+                int fLength = fragment.length();
+                if (fLength == 0) {
+                    continue;
+                }
+                
+                // Add any start/end of string character
+                if (i == 0 && patternStructure.openStart == false) {
+                    // Start-of-string anchored (is not a leading wildcard)
+                    fragment = TOKEN_START_OR_END_CHAR + fragment;
+                }
+                if (patternStructure.openEnd == false && i == patternStructure.fragments.length - 1) {
+                    // End-of-string anchored (is not a trailing wildcard)
+                    fragment = fragment + TOKEN_START_OR_END_CHAR;
+                }
+                
+                if (fragment.length() == numChars) {
+                    TermQuery tq = new TermQuery(new Term(name(), fragment));
+                    bqBuilder.add(new BooleanClause(tq, Occur.MUST));                    
+                } else if (fragment.length() > numChars) {
+                    // Break fragment into multiple Ngrams                
+                    KeywordTokenizer kt = new KeywordTokenizer(256);
+                    kt.setReader(new StringReader(fragment));
+                    TokenFilter filter = new NGramTokenFilter(kt, numChars, numChars, false);
+                    CharTermAttribute termAtt = filter.addAttribute(CharTermAttribute.class);
+                    String lastUnusedToken = null;
+                    try {
+                        filter.reset();
+                        int nextRequiredCoverage = 0;
+                        int charPos = 0;
+                        // minimise number of terms searched - eg for "1234567" and 4grams we only need terms 
+                        // `1234` and `4567` - no need to search for 2345 and 3456
```

Comment:
maybe change the example to use 3 grams since this is the default

Reply:
This is an optimization for phrase queries, but without positions, removing an ngram means that we need to verify all the matches with the doc values ?
I wonder if this would be easier to handle the limit of terms in a custom query ? We can refine the strategy but we should first limit the number of clauses globally, not per fragment. We could also choose to remove wildcard queries in priority if we have more restrictive clauses (term query) and the total number of clauses is beyond the limit. 
Since we're ok with an approximation if the pattern is too big, I wonder if we should aggressively remove leaf clauses if the total number reaches `128` (smaller than max boolean clause) ?